cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/Operand.qll

```diff
@@ -0,0 +1,360 @@
+private import internal.IRInternal
+import Instruction
+import IRBlock
+import cpp
+import semmle.code.cpp.ir.implementation.MemoryAccessKind
+private import semmle.code.cpp.ir.internal.OperandTag
+
+private newtype TOperand =
+  TNonPhiOperand(Instruction instr, OperandTag tag, Instruction defInstr) {
+    defInstr = Construction::getInstructionOperand(instr, tag)
+  } or
+  TPhiOperand(PhiInstruction instr, Instruction defInstr, IRBlock predecessorBlock) {
+    defInstr = Construction::getPhiInstructionOperand(instr, predecessorBlock)
+  }
+
+/**
+ * A source operand of an `Instruction`. The operand represents a value consumed by the instruction.
+ */
+class Operand extends TOperand {
+  string toString() {
+    result = "Operand"
+  }
+
+  /**
+   * Gets the `Instruction` that consumes this operand.
+   */
+  Instruction getInstruction() {
+    none()
+  }
+
+  /**
+   * Gets the `Instruction` whose result is the value of the operand.
+   */
+  Instruction getDefinitionInstruction() {
+    none()
+  }
+
+  /**
+   * Gets a prefix to use when dumping the operand in an operand list.
+   */
+  string getDumpLabel() {
+    result = ""
+  }
+
+  /**
+   * Gets a string describing this operand, suitable for display in IR dumps. This consists of the
+   * result ID of the instruction consumed by the operand, plus a label identifying the operand
+   * kind.
+   *
+   * For example: `this:r3_5`
+   */
+  final string getDumpString() {
+    result = getDumpLabel() + getDefinitionInstruction().getResultId()
+  }
+
+  /**
+   * Get the order in which the operand should be sorted in the operand list.
+   */
+  int getDumpSortOrder() {
+    result = -1
+  }
+
+  /**
+   * Gets the kind of memory access performed by the operand. Holds only for memory operands.
+   */
+  MemoryAccessKind getMemoryAccess() {
+    none()
+  }
+
+  /**
+   * Returns the operand that holds the memory address from which the current operand loads its
+   * value, if any. For example, in `r3 = Load r1, m2`, the result of `getAddressOperand()` for `m2`
+   * is `r1`.
+   */
+  final AddressOperand getAddressOperand() {
+    getMemoryAccess() instanceof IndirectMemoryAccess and
+    result.getInstruction() = getInstruction()
```

Comment:
There's no uniqueness guarantee here. Is that implicitly imposed by `Instruction` construction?

Reply:
There is a uniqueness guarantee, but its enforcement is in the sanity queries. An `Operand`'s identity consists of its owning `Instruction`, its definition `Instruction`, and its `OperandTag`. The `duplicateOperands` sanity query ensures that we do not have two `Operand`s on the same `Instruction` with the same `OperandTag` but different definition definition `Instructions`.